### PR TITLE
Add `rustler_precompiled` and workflow for precompiling NIFs

### PR DESCRIPTION
**Pull request by philss**
This PR contains the changes necessary to use precompiled NIFs.

It will build on every push to the `main` branch and push of tags. But with tags we are also automatically creating a new release that contains all the precompiled NIFs for the most common architectures and operational systems.
After the release, the Hex package can be published as described in the [Precompilation Guide](https://hexdocs.pm/rustler_precompiled/precompilation_guide.html) (only for project maintainers).

The NIF will be downloaded based on the target system (normally the same of the host system, but can be different for example when building for [Nerves](https://www.nerves-project.org/)) and will be verified with a checksum that was generated in the release of the Hex package.

This shouldn't change much of the development flow, so if you are using a pre-release it won't affect you. Once we have a new release, the users of the Hex package will be able to use Explorer without having to install the Rust toolchain.

Even with a released version, it is still possible to force the compilation by setting a system env var:

    EXPLORER_BUILD=true mix compile --force

Or by setting an application env:

    config :rustler_precompiled, :force_build, explorer: true

For more details about `rustler_precompiled`, please [check the documentation](https://hexdocs.pm/rustler_precompiled).

### CODE REVIEW

**Review by cigrainger — APPROVED** [review]
Amazing. I think we should put something in the README re: getting started, but I'm happy to do that in a followup PR.

Thanks @philss! Are you comfortable for this to be merged?

File: mix.exs [hunk 1]
```diff
@@ -50,4 +51,22 @@ defmodule Explorer.MixProject do
...
     ]
   end
+
+  defp package do
+    [
+      files: [
+        "lib",
+        "native",
+        "datasets",
+        # "notebooks", should we include this one?
```
Comment (josevalim) [inline mix.exs:61]: Notebooks are not necessary.

```suggestion
```

File: mix.exs [hunk 2]
```diff
@@ -50,4 +51,22 @@ defmodule Explorer.MixProject do
...
+        "checksum-*.exs",
+        "mix.exs",
+        "README.md",
+        # "CHANGELOG.md",
```
Comment (josevalim) [inline mix.exs:65]:
```suggestion
        "CHANGELOG.md",
```

File: mix.exs [hunk 3]
```diff
@@ -50,4 +51,21 @@ defmodule Explorer.MixProject do
...
+  defp package do
+    [
+      files: [
+        "lib",
+        "native",
+        "datasets",
+        "checksum-*.exs",
+        "mix.exs",
+        "README.md",
+        "CHANGELOG.md",
```
Comment (josevalim) [inline mix.exs:64]:
```suggestion
```

**Review by josevalim — APPROVED** [review]
Beautiful work @philss! This is good to go!

@cigrainger for the README, we only need to note that the release package does not require Rust installed. The `EXPLORER_BUILD` flag that only makes sense once the package is published. Everything else should just work™!

**Comment by philss** [comment]
> I think we should put something in the README re: getting started, but I'm happy to do that in a followup PR.

Oh, please go ahead then. I think the only thing to mention is the ways to force the build when people are using a released version, like José said.

I think it's ready to go! :D
 But let me know if you need help when releasing the package for the first time. The only detail is the checksum file that depends on the "release build". The [Release flow](https://hexdocs.pm/rustler_precompiled/precompilation_guide.html#the-release-flow) explains that in depth.

Thanks, guys!

**Comment by dustinfarris** [comment]
I'm having trouble using this — explorer is still compiling polars

**Comment by dustinfarris** [comment]
possible to get a pre-release then? 😅

**Pull request merged**